### ISSUE TITLE
EitherT doesn't play nice with Task

### ISSUE THREAD

**Issue by nfma** [gcanti/fp-ts#219]
We've stolen the TaskEither example and implemented it for one of our projects. But it suffers from a small issue; failed (left) tasks run twice when chained with other tasks. Below is a test that we've written that fails and proves it:

```javascript
import * as te from "../src/taskEither"

  it("calls first task only once on chain (left)", async () => {
    let called = 0
    const fn = () => { called++ ; return Promise.resolve(1) }
    await te.left(new task.Task(() => fn())).chain((a) => throwUnexpectedCall()).fold(
      (l) => expect(l).toBe(1),
      (r) => throwUnexpectedCall()).run()
    expect(called).toBe(1)
  })
```
I've spent some time looking at this and getting familiar with the code. I'm more familiar with scalaz and it helped that most of the implementations are "borrowed" from scalaz. I've been able to pin-point the problem to the EitherT chain method. The left hand side returns the original task, even after running it and getting it's result back, so in order for the result to be returned at the end that task needs to run again. To avoid that we should build a Task at that point with the result, which we can't do without restricting the type of F in the chain method to be Applicative as well as below:

```javascript
chain<F>(F: Chain<F> & Applicative<F>): EitherT<F>['chain'] {
    return (f, fa) => F.chain(e => e.fold(a => F.of(either.left(a)), a => f(a)), fa)
  }
```

This change fixes the test above but adds a restriction to EitherT.chain. I'm not sure how happy are you to solve the problem this way or if there are better solutions out there that I'm not aware of.

I've got the repo cloned and can submit a PR if you're happy with this solution.

**Comment by gcanti** [comment]
@nfma Thanks for digging into this! So we should require a `Monad` (= Chain + Applicative https://github.com/gcanti/fp-ts/blob/7900f530a1cd5dcefabe7665f1721daf4354cc05/src/Monad.ts#L4)

**Comment by nfma** [comment]
I've collapsed the Applicative and Chain into a Monad as you can see on the PR.

**Comment by gcanti** [comment]
Thanks @nfma, patch released
